File: backend/models/EEGData.js
// models/EEGData.js - EEG Data model
const mongoose = require('mongoose');

const eegDataSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  filename: {
    type: String,
    required: true,
    trim: true
  },
  originalFilename: {
    type: String,
    required: true,
    trim: true
  },
  format: {
    type: String,
    required: true,
    enum: ['edf', 'bdf', 'vhdr', 'set', 'fif', 'cnt', 'npy'],
    lowercase: true
  },
  size: {
    type: Number,
    required: true
  },
  uploadDate: {
    type: Date,
    default: Date.now
  },
  metadata: {
    subject: {
      id: String,
      age: Number,
      gender: String,
      group: String
    },
    session: String,
    task: String,
    acquisition: String,
    channels: Number,
    sampleRate: Number,
    duration: Number
  },
  bidsCompliant: {
    type: Boolean,
    default: false
  },
  tags: [String],
  notes: String,
  svm_analysis: {
    requested: Boolean,
    performed: Boolean,
    in_progress: Boolean,
    result: String,
    confidence: Number,
    performed_at: Date,
    error: String,
    details: mongoose.Schema.Types.Mixed
  },
  dataUrl: String,
  // For storing the EEG data in MongoDB
  // (for small files or when GridFS isn't used)
  data: Buffer,
  // For GridFS reference (for larger files)
  gridFsId: mongoose.Schema.Types.ObjectId
});

// Create indexes for efficient queries
eegDataSchema.index({ userId: 1, uploadDate: -1 });
eegDataSchema.index({ 'metadata.subject.id': 1 });
eegDataSchema.index({ tags: 1 });
eegDataSchema.index({ format: 1 });
eegDataSchema.index({ bidsCompliant: 1 });

// Instance method to request SVM analysis
eegDataSchema.methods.requestAnalysis = function() {
  this.svm_analysis = {
    requested: true,
    performed: false,
    in_progress: false
  };
  return this.save();
};

// Static method to find EEG data for a specific user
eegDataSchema.statics.findByUser = function(userId) {
  return this.find({ userId })
    .sort({ uploadDate: -1 })
    .select('-data'); // Exclude binary data from results
};

// Virtual for file extension
eegDataSchema.virtual('fileExtension').get(function() {
  return this.format ? `.${this.format.toLowerCase()}` : '';
});

const EEGData = mongoose.model('EEGData', eegDataSchema);

module.exports = EEGData;

File: backend-csharp/Models/EegData.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using System.ComponentModel.DataAnnotations;

namespace EegilityApi.Models;

public class EegData
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = string.Empty;

    [Required]
    [BsonRepresentation(BsonType.ObjectId)]
    public string UserId { get; set; } = string.Empty;

    [Required]
    public string Filename { get; set; } = string.Empty;

    [Required]
    public string OriginalFilename { get; set; } = string.Empty;

    [Required]
    public EegFormat Format { get; set; }

    [Required]
    public long Size { get; set; }

    public DateTime UploadDate { get; set; } = DateTime.UtcNow;

    public EegMetadata Metadata { get; set; } = new();

    public bool BidsCompliant { get; set; } = false;

    public List<string> Tags { get; set; } = new();

    public string Notes { get; set; } = string.Empty;

    public AdhdAnalysis? AdhdAnalysis { get; set; }

    public string? DataUrl { get; set; }

    // For GridFS reference (for larger files)
    [BsonRepresentation(BsonType.ObjectId)]
    public string? GridFsId { get; set; }

    // BIDS-specific fields
    public BidsMetadata BidsData { get; set; } = new();

    // Navigation property
    public virtual User? User { get; set; }
}

public enum EegFormat
{
    Edf,
    Bdf,
    Vhdr,
    Set,
    Fif,
    Cnt,
    Npy
}

public class EegMetadata
{
    public SubjectMetadata Subject { get; set; } = new();
    public string Session { get; set; } = string.Empty;
    public string Task { get; set; } = string.Empty;
    public string Acquisition { get; set; } = string.Empty;
    public int Channels { get; set; }
    public double SampleRate { get; set; }
    public double Duration { get; set; }
    public Dictionary<string, object> CustomFields { get; set; } = new();
}

public class SubjectMetadata
{
    public string Id { get; set; } = string.Empty;
    public int? Age { get; set; }
    public string Gender { get; set; } = string.Empty;
    public string Group { get; set; } = string.Empty;
    public string Handedness { get; set; } = string.Empty;
    public List<string> Medications { get; set; } = new();
    public string Diagnosis { get; set; } = string.Empty;
}

public class BidsMetadata
{
    public string Subject { get; set; } = string.Empty;
    public string Session { get; set; } = string.Empty;
    public string Task { get; set; } = string.Empty;
    public string Acquisition { get; set; } = string.Empty;
    public string Run { get; set; } = string.Empty;
    public string Recording { get; set; } = string.Empty;
    public Dictionary<string, object> Participants { get; set; } = new();
    public Dictionary<string, object> TaskMetadata { get; set; } = new();
}

public class AdhdAnalysis
{
    public bool Requested { get; set; } = false;
    public bool Performed { get; set; } = false;
    public bool InProgress { get; set; } = false;
    public string Result { get; set; } = string.Empty;
    public double Confidence { get; set; }
    public DateTime? PerformedAt { get; set; }
    public string? Error { get; set; }
    public AdhdAnalysisDetails? Details { get; set; }
}

public class AdhdAnalysisDetails
{
    public Dictionary<string, double> Probabilities { get; set; } = new();
    public List<string> FeaturesUsed { get; set; } = new();
    public KeyFeatures KeyFeatures { get; set; } = new();
    public string ModelVersion { get; set; } = string.Empty;
}

public class KeyFeatures
{
    public double? ThetaBetaRatio { get; set; }
    public double? FrontalTheta { get; set; }
    public double? CentralBeta { get; set; }
    public double? AlphaActivity { get; set; }
    public Dictionary<string, double> AdditionalFeatures { get; set; } = new();
}

// DTOs
public class EegUploadDto
{
    [Required]
    public IFormFile File { get; set; } = null!;

    public string Notes { get; set; } = string.Empty;
    public List<string> Tags { get; set; } = new();
    
    // Subject metadata
    public string SubjectId { get; set; } = string.Empty;
    public int? SubjectAge { get; set; }
    public string SubjectGender { get; set; } = string.Empty;
    public string SubjectGroup { get; set; } = string.Empty;
    
    // Session metadata
    public string Session { get; set; } = string.Empty;
    public string Task { get; set; } = string.Empty;
    public string Acquisition { get; set; } = string.Empty;
}

public class EegDataResponseDto
{
    public string Id { get; set; } = string.Empty;
    public string Filename { get; set; } = string.Empty;
    public string OriginalFilename { get; set; } = string.Empty;
    public string Format { get; set; } = string.Empty;
    public long Size { get; set; }
    public DateTime UploadDate { get; set; }
    public EegMetadata Metadata { get; set; } = new();
    public bool BidsCompliant { get; set; }
    public List<string> Tags { get; set; } = new();
    public string Notes { get; set; } = string.Empty;
    public AdhdAnalysis? AdhdAnalysis { get; set; }
    public string? DataUrl { get; set; }
}

public class AdhdAnalysisRequestDto
{
    [Required]
    public string EegDataId { get; set; } = string.Empty;
}